distributed_shampoo/fsdp_shampoo.py

```diff
@@ -363,33 +372,86 @@ def _initialize_preconditioners_and_steps(self):
+                        grafting_epsilon=self._grafting_epsilon,
+                        use_protected_eigh=self._use_protected_eigh,
+                        use_dtensor=self._use_dtensor,
+                    )
+                elif self._convex_shape_recovery == ConvexShapeRecoveryMethod.COMM:
+                    state[PRECONDITIONERS] = CommunicationShampooPreconditioner(
+                        p,
+                        self._param_metadata[p],
+                        large_dim_method=self._large_dim_method,
+                        beta2=group[BETAS][1],
+                        epsilon=group[EPSILON],
+                        exponent_override=self._exponent_override,
+                        exponent_multiplier=self._exponent_multiplier,
+                        use_bias_correction=self._use_bias_correction,
+                        max_preconditioner_dim=self._max_preconditioner_dim,
+                        dtype=self._preconditioner_dtype,
+                        idx=idx,
+                        use_merge_dims=self._use_merge_dims,
+                        start_preconditioning_step=self._start_preconditioning_step,
+                        grafting_type=self._grafting_type,
+                        grafting_beta2=self._grafting_beta2,
+                        grafting_epsilon=self._grafting_epsilon,
+                        use_protected_eigh=self._use_protected_eigh,
+                        use_dtensor=self._use_dtensor,
+                        # TODO: can do more complex assignment here
+                        left_comm=CommunicationType.NONE
+                        if group_rank == 0
+                        else CommunicationType.RECV,
+                        right_comm=CommunicationType.NONE
+                        if group_rank == dist.get_world_size() - 1
+                        else CommunicationType.SEND,
+                    )
+                else:
+                    raise NotImplementedError("invalid convex shape recovery method")
 
                 # Count parameters from preconditioners for logging purposes.
                 self._parameter_count += state[PRECONDITIONERS].parameter_count
 
         # Logs total number of parameters for optimizer.
         logger.info(f"Total Parameter Count: {self._parameter_count}")
 
+    @torch.no_grad()
+    def _send_grad(self, direction: CommunicationDirection):
+        # Get communication ops from all of the preconditioners.
+        ops = []
+        for group in self.param_groups:
+            for p in group[PARAMS]:
+                # skip parameters not on worker
+                if p.numel() == 0 or p.grad is None:
+                    continue
+                state = self.state[p]
+
+                if direction == CommunicationDirection.FORWARD:
+                    ops.extend(state[PRECONDITIONERS].get_forward_ops(p.grad))
+                else:  # backward
+                    ops.extend(state[PRECONDITIONERS].get_backward_ops())
```

Comment:
I don't know if there are any other possible `CommunicationDirection` options; perhaps we should change this just in case?
```suggestion
                elif direction == CommunicationDirection.BACKWARD:
                    ops.extend(state[PRECONDITIONERS].get_backward_ops())
                else:
                    logger.warning(f"Communication direction {CommunicationDirection} is not valid! Continuing...")
```